### PR TITLE
Hotcue Quantize fix 

### PR DESCRIPTION
**Pull request by daschuer**
This fixes:
https://bugs.launchpad.net/mixxx/+bug/1777429
A test is in place.

### CODE REVIEW

File: src/engine/cuecontrol.cpp [hunk 1]
```diff
@@ -481,23 +481,23 @@ void CueControl::hotcueActivatePreview(HotcueControl* pControl, double v) {
     if (v) {
         if (pCue && pCue->getPosition() != -1) {
             m_iCurrentlyPreviewingHotcues++;
-            int iPosition = pCue->getPosition();
```
Comment (Pegasus-RPG) [inline src/engine/cuecontrol.cpp]: Was there a reason we used integer here in the past? Perhaps double is too precise?

File: src/engine/enginebuffer.h [hunk 2]
```diff
@@ -91,9 +91,17 @@ class EngineBuffer : public EngineObject {
   public:
     enum SeekRequest {
         SEEK_NONE = 0x00,
-        SEEK_PHASE = 0x01,
-        SEEK_EXACT = 0x02,
-        SEEK_STANDARD = 0x03, // = (SEEK_EXACT | SEEK_PHASE)
+        SEEK_PHASE = 0x01, // This be set to force seek into phase
```
Comment (Pegasus-RPG) [inline src/engine/enginebuffer.h]: "This **is** set..."

File: src/engine/enginebuffer.h [hunk 3]
```diff
@@ -91,9 +91,17 @@ class EngineBuffer : public EngineObject {
...
+        SEEK_EXACT = 0x02, // This is used to seek to position regardless of
+                           // Quantize enabled
```
Comment (Pegasus-RPG) [inline src/engine/enginebuffer.h]: "regardless of if Quantize is enabled"

File: src/engine/enginebuffer.h (hunk 2 above)
Comment (Pegasus-RPG) [inline src/engine/enginebuffer.h]: "This is set to force an in-phase seek."

File: src/engine/cuecontrol.cpp [hunk 4]
```diff
@@ -481,23 +481,23 @@ void CueControl::hotcueActivatePreview(HotcueControl* pControl, double v) {
...
+            double position = pCue->getPosition();
```
Comment (Pegasus-RPG) [inline src/engine/cuecontrol.cpp]: Was there a reason we used integer here in the past? Perhaps double is too precise? (Are there any clues in the commit history on this line?)

File: src/engine/enginebuffer.h [hunk 5]
```diff
@@ -91,9 +91,17 @@ class EngineBuffer : public EngineObject {
...
+        SEEK_EXACT_PHASE = 0x03, // This is an artificial state the happens if
```
Comment (Pegasus-RPG) [inline src/engine/enginebuffer.h]: "...state **that** happens..."

File: src/engine/enginebuffer.h [hunk 6]
```diff
@@ -91,9 +91,17 @@ class EngineBuffer : public EngineObject {
...
+                                 // an exact seek and a phase seek are sheduled
```
Comment (Pegasus-RPG) [inline src/engine/enginebuffer.h]: "scheduled"

File: src/engine/enginebuffer.h [hunk 7]
```diff
@@ -91,9 +91,17 @@ class EngineBuffer : public EngineObject {
...
+                                 // at the same time.
+        SEEK_STANDARD = 0x04, // This is used to seek exact or phase depending
+                              // of Quantize enabled
```
Comment (Pegasus-RPG) [inline src/engine/enginebuffer.h]: "This seeks to the exact position if Quantize is disabled or performs an in-phase seek if it is enabled."

File: src/engine/enginebuffer.h [hunk 8]
```diff
@@ -91,9 +91,17 @@ class EngineBuffer : public EngineObject {
...
+        SEEK_STANDARD_PHASE = 0x05, // This is an artificial state the happens if
+                                    // an standard seek and a phase seek are sheduled
+                                    // at the same time.
```
Comment (Pegasus-RPG) [inline src/engine/enginebuffer.h]: "...state **that** happens if **a** standard...are **scheduled**..."

File: src/test/enginesynctest.cpp [hunk 9]
```diff
@@ -1497,3 +1497,48 @@ TEST_F(EngineSyncTest, QuantizeImpliesSyncPhase) {
...
+    ProcessBuffer();
+    pHotCueActivate->set(0.0);
+    ProcessBuffer();
+
+    ControlObject::set(ConfigKey(m_sGroup2, "quantize"), 0.0);
+    // preview a hot cue without quantize
+    pHotCueActivate->set(1.0);
+    ProcessBuffer();
+
+    ASSERT_DOUBLE_EQ(0.019349962207105064, ControlObject::get(ConfigKey(m_sGroup2, "beat_distance")));
```
Comment (Pegasus-RPG) [inline src/test/enginesynctest.cpp]: Will this always be this exact or should this rather be a small range of valid values?

File: src/test/enginesynctest.cpp [hunk 10]
```diff
@@ -1497,3 +1497,48 @@ TEST_F(EngineSyncTest, QuantizeImpliesSyncPhase) {
...
+
+    pHotCueActivate->set(0.0);
+    ProcessBuffer();
+
+    ControlObject::set(ConfigKey(m_sGroup2, "quantize"), 1.0);
+    // preview a hot cue without quantize
```
Comment (Pegasus-RPG) [inline src/test/enginesynctest.cpp]: "**with** quantize."

**Review by Pegasus-RPG — CHANGES_REQUESTED** [review]

File: src/test/enginesynctest.cpp (hunk 9 above)
Comment (daschuer) [reply src/test/enginesynctest.cpp]: ASSERT_DOUBLE_EQ already uses an elide. We will see in the CI if this requires a range.

**Comment by daschuer** [comment]
Done!

**Comment by daschuer** [comment]
Ping!

**Comment by Pegasus-RPG** [comment]
What benefit is there in adding the two new artificial states?

**Comment by daschuer** [comment]
These states happens in case of race conditions of two seek commands. It should normally not happen, but it is correct to handle them.

**Comment by Pegasus-RPG** [comment]
Can you expand the tests to test all possible cases? (So we get a sensible result from all and make sure there's no crash.)

**Comment by Pegasus-RPG** [comment]
For all tests, you should probably also test setting hot cues before and after the current play position to catch any future sign errors.

**Comment by daschuer** [comment]
> Can you expand the tests to test all possible cases? (So we get a sensible result from all and make sure there's no crash.)

IMHO all significant cases are covered by the tests. If you see a missing test case I am happy to add it. The previous play position is not used for caculate the new cue position, so a jump before and behind the track makes no difference.
It would be interesting to run a code coverage tool to identify blind spots though.

**Comment by daschuer** [comment]
Merge? This fixes an anoying 2.1.1 regression for some user's mayor use cases.

File: src/engine/enginebuffer.h [hunk 11]
```diff
@@ -91,9 +91,17 @@ class EngineBuffer : public EngineObject {
   public:
     enum SeekRequest {
         SEEK_NONE = 0x00,
-        SEEK_PHASE = 0x01,
-        SEEK_EXACT = 0x02,
-        SEEK_STANDARD = 0x03, // = (SEEK_EXACT | SEEK_PHASE)
+        SEEK_PHASE = 0x01, // This is set to force an in-phase seek.
+        SEEK_EXACT = 0x02, // This is used to seek to position regardless of
+                           // if Quantize is enabled.
+        SEEK_EXACT_PHASE = 0x03, // This is an artificial state that happens if
```
Comment (uklotzde) [inline src/engine/enginebuffer.h]: SEEK_EXACT_PHASE = SEEK_EXACT | SEEK_PHASE,
SEEK_STANDARD_PHASE = SEEK_STANDARD | SEEK_PHASE,
would be more explicit for the derived literals.

File: src/test/enginesynctest.cpp (hunk 9 above)
Comment (uklotzde) [inline src/test/enginesynctest.cpp]: Can we add a remark on how this value has been calculated? Or even better replace the magic number with the corresponding expression?

**Comment by daschuer** [comment]
These values were determined experimentally.
Done!

**Comment by daschuer** [comment]
Merge?

**Comment by daschuer** [comment]
@Pegasus-RPG: can we merge this now?

**Comment by Pegasus-RPG** [comment]
LGTM. (Sorry, I've been busy with non-Mixxx things lately.)

**Comment by Pegasus-RPG** [comment]
Those experimentally-determined values still make me nervous, so we'll need to keep this in mind if new bugs come up related to this change.

**Pull request merged**